### PR TITLE
[megatron, ckpt] refactor: optimize mcore ckpt manager impl, checkpoint yaml config APIs

### PR DESCRIPTION
**Pull request by ETOgaosion**
### What does this PR do?

Refactoring Megatron checkpoint menager to avoid complex logics, following the way mentioned in https://github.com/verl-project/verl/issues/5630

# Problem

Now the mcore checkpoint manager has multiple backends: `mbridge, dist_checkpointing, verl original implemenation`, and there are 2 model related `save_contents` APIs: `model, hf_model`, which is complex in implemenation.

Also, the API implementation in codes is complex, `save_contents` are transformed with `should_save_xxx` judgement and the backends converted to `use_xxx` judgement. These naming is confusing and makes users difficult to understand saving logics.

# Refactor: Megatron checkpoint layout + checkpoint manager

## 1. New checkpoint structure (layout schema v2)

Follow the style of FSDP checkpoint manager, save by contents:

```
checkpoints/${project}/${experiment}/global_step_${i}/${role}/
├── ckpt_contents.json          # manifest (written last = save is complete)
├── transformer_config.json     # rank-0, when 'extra' is saved
├── model/
│   ├── huggingface/            # mbridge: HF weights + config + tokenizer
│   └── dist_ckpt/              # mbridge off: model shards / mbridge on + PEFT: adapter shards
├── optimizer/
│   └── dist_ckpt/              # optimizer + lr_scheduler shards
└── extra/
    └── dist_ckpt/              # rng_state shards
```

Old (v1) layout, for reference:

```
${role}/
├── dist_ckpt/        # optimizer + rng + (maybe) model shards, all mixed
├── huggingface/      # optional
└── transformer_config.json
```

## 2. New input behavior

- Backend flag renamed: `use_dist_checkpointing` → `use_mbridge`.
  - `use_dist_checkpointing` is still accepted as a legacy alias (`use_mbridge = not use_dist_checkpointing`).
- `save_contents` × backend matrix (validated up front in `__init__`):

  | Backend              | `save_contents`   | Behaviour                                         |
  |----------------------|-------------------|---------------------------------------------------|
  | `mbridge` (default)  | `model`           | HF save into `model/huggingface/`                 |
  | `mbridge`            | `hf_model`        | HF save into `model/huggingface/`                 |
  | `mbridge`            | both              | Same save, deduplicated (written once)            |
  | `dist_checkpoint`    | `model`           | Sharded save into `model/dist_ckpt/`              |
  | `dist_checkpoint`    | `hf_model`        | **Error** — only mbridge supports `hf_model`      |
  | both disabled        | any               | **Error** — at least one model backend required   |

- `optimizer` and `extra` always go through `dist_checkpointing`, regardless of backend.
- Loading a v1 checkpoint raises a clear error with a pointer to the migration script.

## 3. Old → new checkpoint converter

`scripts/migrate_megatron_checkpoint_layout.py`:

```bash
# Single role directory:
python scripts/migrate_megatron_checkpoint_layout.py \
    --checkpoint /path/to/global_step_N/actor

# Every step under a run:
python scripts/migrate_megatron_checkpoint_layout.py \
    --checkpoint-root /path/to/run --all-steps
```

- `--mode hardlink` (default) / `symlink` / `copy`.
- Idempotent (no-op on a v2 checkpoint).
- Works because `dist_checkpointing.load` is key-driven — it ignores unexpected entries, so the migrator can point `model/dist_ckpt/`, `optimizer/dist_ckpt/`, `extra/dist_ckpt/` at the same original `dist_ckpt/` data.

## 4. `MegatronCheckpointManager` structure

The manager is now organised into four layers:

- **util** — path helpers for every component of the v2 layout (`model/`, `model/dist_ckpt/`, `model/huggingface/`, `optimizer/dist_ckpt/`, `extra/dist_ckpt/`, `transformer_config.json`, `ckpt_contents.json`), plus detectors for the legacy v1 layout.
- **build** — composable builders for the sharded state dict (model / optimizer / extra), plus the `ckpt_contents.json` manifest payload. Save and load paths now only build the pieces they actually need, instead of one monolithic dict.
- **load** — validates layout (v1 → migration hint), reads the manifest, dispatches to the right backend (mbridge / dist_ckpt / PEFT).
- **save** — dispatches to the right backend per component, writes HF/tokenizer/transformer_config side files, then writes the manifest last via atomic rename so its presence signals a complete checkpoint.

cc @ISEEKYAN @vermouth1992 @PeterSH6 @wuxibin89 @khazic @tongyx361 

### Checklist Before Starting

- [ ] Search for similar PRs. Paste at least one query link here: ...
- [ ] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `fully_async`, `one_step_off`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

Can continuous training:

image

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [ ] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [ ] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

File: docs/ascend_tutorial/features/ascend_backend_features.md [hunk 1]
```diff
@@ -249,7 +249,7 @@ class MindSpeedFeature:
 | verl参数 | 简介|
 | --- | --- |
 | `actor_rollout_ref.actor.megatron.optimizer_offload` |是否卸载模型优化器到CPU，默认值为False|
-| `actor_rollout_ref.actor.megatron.use_mbridge` |是否使用mbridge进行权重转换|
+| `actor_rollout_ref.actor.megatron.use_mbridge` |是否使用mbridge进行权重转换；为 True（默认）时模型权重以HuggingFace 格式通过 mbridge 保存/加载，且 `save_contents` 中的 `model` 与 `hf_model` 会去重只保存一次；为 False 时使用 Megatron `dist_checkpointing` 保存模型分片，此时 `save_contents` 不允许包含 `hf_model`|
```
Comment (ISEEKYAN) [inline docs/ascend_tutorial/features/ascend_backend_features.md]: use_mbridge's original meaning is to choose the way of integrating megatron, if False, then the config_converter way.
But now we have removed the config_converter and the `use_mbridge` is no longer valid. I don't recommend reuse the argument name to indicate `whether to save HF format weights`, it is better to add a new option such as `save_HF_weights`

File: verl/workers/engine/megatron/transformer_impl.py [hunk 2]
```diff
@@ -419,10 +419,10 @@ def initialize(self):
             optimizer_scheduler=self.lr_scheduler,
             use_distributed_optimizer=self.engine_config.use_distributed_optimizer,
             use_checkpoint_opt_param_scheduler=self.optimizer_config.use_checkpoint_opt_param_scheduler,
+            use_mbridge=not self.engine_config.use_dist_checkpointing,
```
Comment (ISEEKYAN) [inline verl/workers/engine/megatron/transformer_impl.py]: better to name it as something like `save_HF_weights`

File: docs/ascend_tutorial/features/ascend_backend_features.md (hunk 1 above)
Comment (ETOgaosion) [reply docs/ascend_tutorial/features/ascend_backend_features.md]: Now we are trying to align with base_ckpt_manager API: should_save_xxx

File: verl/workers/engine/megatron/transformer_impl.py (hunk 2 above)
Comment (ETOgaosion) [reply verl/workers/engine/megatron/transformer_impl.py]: Will remove this, and negative use_dist_checkpointing inside ckpt manager init function

**Comment by ETOgaosion** [comment]
@Kirrito-k423 Please also help review~

**Review by wuxibin89 — APPROVED** [review]
